test/test_autograd.py

```diff
@@ -855,6 +855,18 @@ def hook(grad):
             self.assertEqual(x.grad, expected_grad)
             self.assertIsNone(x_list[i].grad)
 
+    def test_materialized_grad(self):
+        x = torch.tensor(0.5, requires_grad=True)
+        a = torch.tensor(1.0, requires_grad=True)
+        y = x * a
+        dydx = torch.autograd.grad(y, x, create_graph=True)
+        d2ydx2_none = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True)
+        d2ydx2 = torch.autograd.grad(dydx, x, create_graph=True, allow_unused=True, materialized_grad=True)
+        d3ydx3 = torch.autograd.grad(d2ydx2, x, allow_unused=True, materialized_grad=True)
+        self.assertIsNone(d2ydx2_none[0])
+        self.assertEqual(d2ydx2[0].item(), 0)
+        self.assertEqual(d3ydx3[0].item(), 0)
+
```

Comment:
Could we add some more test cases for the new interactions between allow_unused and materialize_grad?:
- user passing in allow_unused=False explicitly and materialize_grad=True should error
- if user does not pass in allow_unused, and allow_unused should take value of materialize_grad